src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala

```diff
@@ -549,7 +549,8 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
       val needsConstructor = (
            !sawPrivateConstructor
         && !(instanceScope containsName nme.CONSTRUCTOR)
-        && ((sflags & INTERFACE) == 0 || (sflags | JAVA_ANNOTATION) != 0)
+        && ((sflags & INTERFACE) == 0 || (sflags & JAVA_ANNOTATION) != 0)
+        && !file.name.endsWith(".sig")
```

Comment:
When do we actually need to add the synthetic constructor, besides (apparently) Java annotations? Could we change the entire thing to `val needsConstructor = sflags & JAVA_ANNOTATION) != 0`?

Reply:
Thanks, I was waiting for a smart person to explain it to me.

Reply:
I meant explain what the constructor construction was for. But I see now, yes, it's just for annotations which IIRC are automatically `new Thing`. I'll experiment with the simplication. And understand why the bad flag test did not matter. Or I guess it would just always supply the ctor.